PSATD Single precision

MaxThevenet opened the pull request:
This PR generalizes spectral solver to single precision. Accuracy was tested on CPU. It contains:
- some `_rt` as cuFFT gave compilation errors with operations `double * Complex<float>`
- calls to single-precision FFT functions for cuFFT (`R2C` instead of `D2Z` etc.)
- calls to single-precision FFT functions for FFTW (`fftwf_...` instead of `fftw_`)
- deleted file `WarpX_ComplexForFFT.H` as it duplicated `Utils/WarpX_Complex.H`

RevathiJambunathan commented:
@MaxThevenet Thank you for this PR. Does changing double to single precision for cuFFT on the GPU change the result at all? and what was the improvement in time?

RevathiJambunathan commented on Source/FieldSolver/SpectralSolver/SpectralFieldData.H:
```diff
@@ -8,7 +8,7 @@
 #define WARPX_SPECTRAL_FIELD_DATA_H_
 
-#include "WarpX_ComplexForFFT.H"
+#include "Utils/WarpX_Complex.H"
```
👍

RevathiJambunathan commented on Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp:
```diff
@@ -73,28 +82,44 @@ SpectralFieldData::SpectralFieldData( const amrex::BoxArray& realspace_ba,
+                              fft_size[1],fft_size[0], CUFFT_R2C);
+#    else
                               fft_size[1],fft_size[0], CUFFT_D2Z);
+#    endif
```
I think this can be
`result = cufftPlan3d( &forward_plan[mfi], fft_size[2], fft_size[1], fft_size[0],
#ifdef AMREX_USE_FLOAT
                                  CUFFT_R2C);
#else 
                                  CUFFT_D2Z);
#endif `

RevathiJambunathan replied:
Also, @MaxThevenet , thanks for teaching me how to write this correctly in the review :)

RevathiJambunathan commented on Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp:
```diff
@@ -73,28 +82,44 @@ SpectralFieldData::SpectralFieldData( const amrex::BoxArray& realspace_ba,
+                              fft_size[1],fft_size[0], CUFFT_C2R);
+#    else
+                              fft_size[1],fft_size[0], CUFFT_Z2D);
+#    endif
```
Same as above. Please can you add only the specific float/versus/double part within the #ifdef region.

MaxThevenet commented on Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp:
```diff
@@ -73,28 +82,44 @@ SpectralFieldData::SpectralFieldData( const amrex::BoxArray& realspace_ba,
+                              fft_size[0], CUFFT_C2R);
+#    else
+                              fft_size[0], CUFFT_Z2D);
+#    endif
```
```suggestion
                              fft_size[0],
#    ifdef AMREX_USE_FLOAT
                              CUFFT_C2R);
#    else
                              CUFFT_Z2D);
#    endif
```

MaxThevenet commented on Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp:
```diff
@@ -106,21 +131,45 @@ SpectralFieldData::SpectralFieldData( const amrex::BoxArray& realspace_ba,
         forward_plan[mfi] =
             // Swap dimensions: AMReX FAB are Fortran-order but FFTW is C-order
 #if (AMREX_SPACEDIM == 3)
+#    ifdef AMREX_USE_FLOAT
```
```suggestion
```

MaxThevenet commented:
Thanks for your review, I implemented the changes + some cleaning in the last few commits. Also added a test, so let's see if tests pass before merging.
